#pipeline with erne (http://erne.sourceforge.net/)
#and bismark (http://www.bioinformatics.babraham.ac.uk/projects/bismark/) using bowtie 2 (http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)

#edit with the actual fasta files to be used in the experiments

wd=/home/nicola/workspace/datasets/alignments/bw-erne/bs_simulated

fasta=/home/nicola/workspace/datasets/genomes/50000/50000.fasta
#fasta=/home/nicola/workspace/datasets/genomes/hg19/ucsc.hg19.fasta

temp_dir=~/workspace/datasets/alignments/bismark/temp/ #bismark temp folder

#genome_folder=~/workspace/datasets/genomes/hg19/bismark/
genome_folder=~/workspace/datasets/genomes/50000/bismark/

#erne_ref=/home/nicola/workspace/datasets/genomes/hg19/ucsc.hg19.ebm
erne_ref=/home/nicola/workspace/datasets/genomes/50000/50000.ebm

# 1) simulate dataset
#rm wd/*

./testcase-bs-aligner-default.sh $fasta $wd


# 2) urn erne

#align with the BS-aligner erne-bs5
erne-bs5 --reference $erne_ref --query1 $wd/query1.fq --query2 $wd/query2.fq --output $wd/erne_bs5_out.bam

#call methylation with the caller erne-meth, generating methylation annotations in bismark format 
erne-meth --fasta $fasta --input $wd/erne_bs5_out.bam --output-prefix $wd/erne_meth_out --annotations-bismark


# 3) run bismark

#align with the BS-aligner bismark
bismark --bowtie2 $genome_folder -1 $wd/query1.fq -2 $wd/query2.fq -o $wd --temp_dir $temp_dir -p 2

samfile=$wd/query1.fq_bismark_bt2_pe.sam

#call methylation with the bismark caller, generating methylation annotations in bismark format 
bismark_methylation_extractor -p --output $wd --gzip --cytosine_report --CX --genome_folder $genome_folder $samfile

#remove all useless files generated by bismark
rm $wd/query1.fq_bismark_bt2_pe.bedGraph $wd/query1.fq_bismark_bt2_pe.bismark.cov $wd/*.txt.gz $wd/query1.fq_bismark_bt2_pe.M-bias.txt


# 4) print stats

#count number of calls that are 20% from the correct value (ERNE)
printf "ERNE results:\n"
./compare_meth_annotations.sh $wd/erne_meth_out_bismark.bed $wd/ 0.2

printf "\nBISMARK results:\n"
./compare_meth_annotations.sh $wd/query1.fq_bismark_bt2_pe.CX_report.txt $wd/ 0.2








